### PR TITLE
API: handle NaN as min/max stats in evaluators

### LINKED ISSUE #1761: Update evaluators to consider NaN for evaluating upper/lower bounds
See [discussion and the test case](https://github.com/apache/iceberg/pull/1641/files#r521660283), due to the comparison implementation of ORC stats, for float/double columns in ORC files, if the first value in a file is NaN, metrics of this file will report NaN for both upper and lower bound despite that the column could contain non-NaN data. This means that today we may unexpectedly skip including an ORC file for predicates that utilize bounds. We need to update evaluators to consider this edge case.

### PR DESCRIPTION
**Pull request by yyanyy**
This PR is based on the behavior we observed in [here](https://github.com/apache/iceberg/blob/master/core/src/test/java/org/apache/iceberg/TestMetrics.java#L362-L417):
- NaN is always populated as max when present in Parquet, and both max and min for ORC if the first row of the file is NaN
- when lower bound is NaN, upper bound will also always be NaN
- in Java, NaN is considered to be greater than all numbers in comparison

This fixes #1761.

### CODE REVIEW

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java [hunk 1]
```diff
@@ -204,15 +210,20 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
...
         return ROWS_CANNOT_MATCH;
       }
 
       if (lowerBounds != null && lowerBounds.containsKey(id)) {
         T lower = Conversions.fromByteBuffer(ref.type(), lowerBounds.get(id));
 
         int cmp = lit.comparator().compare(lower, lit.value());
-        if (cmp > 0) {
+
+        // Due to the comparison implementation of ORC stats, for float/double columns in ORC files,
```
Comment (jackye1995) [inline api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java:222]: Thanks for noticing this. Instead of having this paragraph everywhere, can we abstract this to a method like `checkNaNLowerBound`, and make this the documentation of that method?

Comment (yyanyy) [reply api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java:222]: Thanks for the quick review! Yeah I do realize that repeating the same comment over and over again is a bit annoying, but I wasn't sure where the right balance is. Since I'm hoping to check for `isNaN` after comparing to avoid unnecessary checking in `lt`/`ltEq`, the only thing I can abstract out is `NaNUtil.isNaN(lower)`, that we are essentially wrapping around a wrapper; and also I guess that might not help much with readability since the actual explanation in this case will be outside of the logic flow here, so the reader will have to jump around to understand the full intention. Maybe we can shorten this comment everywhere and have the full version at the start of the class? Do you/other people have any suggestion?

Comment (jackye1995) [reply api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java:222]: Sorry I did not see the comment, I think it is valuable enough to abstract out the method given the amount of comments, and have something like:

```
/*
 * all the comments...
 */
private boolean isLowerBoundNaN() {
  return NaNUtil.isNaN(lower);
}

```

I am not good at naming, you can probably come up with a better name...

**Comment by jbapple** [comment]
I think in Java NaN is neither greater then, nor equal to, nor less than any numbers. That’s the IEEE754 semantics, at least.

**Comment by yyanyy** [comment]
> I think in Java NaN is neither greater then, nor equal to, nor less than any numbers. That’s the IEEE754 semantics, at least.

Thanks for the interests in this PR! I think in Java, in comparisons NaN is handled as greater than all floating point numbers including infinity, but you are right NaN shouldn't normally be comparable.

**Comment by jbapple** [comment]
Odd, maybe we're trying different things. What does this print for you?

```
public class Comp {
  public static void main(String[] args) {
     System.out.println(Double.NaN > 0.0);
  }
}
```

I'm on OpenJDK 1.8, and this prints `false` for me.

**Comment by yyanyy** [comment]
> Odd, maybe we're trying different things. What does this print for you?
> 
> ```
> public class Comp {
>   public static void main(String[] args) {
>      System.out.println(Double.NaN > 0.0);
>   }
> }
> ```
> 
> I'm on OpenJDK 1.8, and this prints `false` for me.

Yeah this prints `false` for me too, as well as `Double.NaN == Double.NaN`. I think we workaround this problem in evaluators by using a [`Comparator.naturalOrder()`](https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/expressions/Literals.java#L156) instead of direct operator comparison, as the following code prints the behavior that NaN is considered larger than positive infinity as described in [here](https://docs.oracle.com/javase/8/docs/api/java/lang/Double.html#compareTo-java.lang.Double-)(by returning -1, 1, 1, -1, 0 respectively)

```
    Comparator comparator = Comparator.naturalOrder();
    System.out.println(comparator.compare(1.2F, Float.NaN));
    System.out.println(comparator.compare(Float.NaN, 1.2F));
    System.out.println(comparator.compare(Float.NaN, Float.POSITIVE_INFINITY));
    System.out.println(comparator.compare(Float.POSITIVE_INFINITY, Float.NaN));
    System.out.println(comparator.compare(Float.NaN, Float.NaN));
```

**Comment by jackye1995** [comment]
@jbapple @yyanyy this blog has a good summary: https://blog.csdn.net/jierui001/article/details/3278382

**Comment by jbapple** [comment]
LGTM. One note I'll make: not all languages treat NaN like Java. For instance, in C++, there are roughly 2^52 different NaN values, and the totalOrder predicate sorts negative NaNs before negative infinity. As such, you could have columns with a NaN lower bound and a numeric upper bound. I'm not sure what the latest draft Iceberg specification says about this.

**Comment by jackye1995** [comment]
> not all languages treat NaN like Java. For instance, in C++, there are roughly 2^52 different NaN values, and the totalOrder predicate sorts negative NaNs before negative infinity.

Yeah that's why I was asking about the spec in #2055. I think What you describe is consistent with the current plan of -NaN < -Infinity < -value < -0 < 0 < value < Infinity < NaN.

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java (hunk 1 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java:222]: I don't that there is a need for an extra method that has just one methdo call. I'd probably do it like this:

```java
        T lower = Conversions.fromByteBuffer(ref.type(), lowerBounds.get(id));
        if (NaNUtil.isNaN(lower)) {
          // NaN indicates unreliable bounds. See the InclusiveMetricsEvaluator docs for more.
          return ROWS_MIGHT_MATCH;
        }

        int cmp = lit.comparator().compare(lower, lit.value());
        if (cmp > 0) {
          return ROWS_CANNOT_MATCH;
        }
```

The docs would go in the javadoc for the whole class, and each NaN check could simply refer back to it. I also moved the NaN check above the comparison to keep the logic simple: if the value is NaN, the bound is invalid.

File: api/src/test/java/org/apache/iceberg/expressions/TestMetricsEvaluatorsNaNHandling.java [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+        SCHEMA, Expressions.in("some_nan_correct_bounds", 1F, 10F, 30F)).eval(FILE);
+    Assert.assertTrue("Should match: 10 within bounds", shouldRead);
+
+    shouldRead = new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.in("some_nan_correct_bounds", 1F)).eval(FILE);
+    Assert.assertFalse("Should not match: 1 not within bounds", shouldRead);
+
+    shouldRead = new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.in("some_nan_correct_bounds", 30F)).eval(FILE);
+    Assert.assertFalse("Should not match: 30 not within bounds", shouldRead);
```
Comment (rdblue) [inline api/src/test/java/org/apache/iceberg/expressions/TestMetricsEvaluatorsNaNHandling.java:230]: Nit: 1F and 30F could be combined to one case, `Expressions.in("some_nan_correct_bounds", 1F, 30F)`.

File: api/src/test/java/org/apache/iceberg/expressions/TestMetricsEvaluatorsNaNHandling.java [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+      Assert.assertFalse("Should not match: 10 is less than upper bound", shouldRead);
+
+      shouldRead = new StrictMetricsEvaluator(SCHEMA, func.apply("min_max_nan", 1F)).eval(FILE);
+      Assert.assertFalse("Should not match: no visibility", shouldRead);
+
+      shouldRead = new StrictMetricsEvaluator(SCHEMA, func.apply("all_nan_null_bounds", 1D)).eval(FILE);
+      Assert.assertFalse("Should not match: all nan column doesn't contain number", shouldRead);
+
+      shouldRead = new StrictMetricsEvaluator(SCHEMA, func.apply("some_nan_correct_bounds", 10F)).eval(FILE);
+      Assert.assertFalse("Should not match: nan value exists", shouldRead);
```
Comment (rdblue) [inline api/src/test/java/org/apache/iceberg/expressions/TestMetricsEvaluatorsNaNHandling.java:270]: This case would return `false` anyway because the correct upper bound is > 10. I think you should change this test case to 30F so that it matches the whole range, but does not match because there are some NaN values.

**Comment by rdblue** [comment]
@yyanyy, the tests look good other than one case that I mentioned. I also commented on the implementation thread, but overall I think this is correct. Nice work!

**Comment by kbendick** [comment]
For ORC metrics, which currently don't populate `nanValueCounts` in the metrics and for which it seems difficult to populate `nanValueCounts`, will this still work when we create `Metrics` using the non-deprecated constructors that require passing `nanValueCounts`? This is in reference to this issue for removing the `Metrics` constructors that don't take in `nanValueCounts` maps: https://github.com/apache/iceberg/issues/2044

I suppose when we go to deprecate the old `Metrics` constructors, we can simply pass `null` into the `Metrics` constructor for `nanValueCounts` with ORG in order to 

Based off of my understanding of ORC's `ColumnStatistics` and the notes in this PR about `upperBound ` and `lowerBound` being `NaN` in ORC if _any_ value in the file is `NaN`. To me, I fail to see how we can get the count of NaN values in the following code block (though we can always continue to pass in `null` to match the existing behavior, provided the various metrics evaluators are aware that `null` for that Map entry for nullValueCounts doesnt convey much meaning for ORC).

The relevant code for populating ORC metrics is here: https://github.com/apache/iceberg/blob/29915e3522f9a808420a2e16fb1c2cf3ae165a74/orc/src/main/java/org/apache/iceberg/orc/OrcMetrics.java#L100-L172

**Comment by yyanyy** [comment]
> For ORC metrics, which currently don't populate `nanValueCounts` in the metrics and for which it seems difficult to populate `nanValueCounts`, will this still work when we create `Metrics` using the non-deprecated constructors that require passing `nanValueCounts`? This is in reference to this issue for removing the `Metrics` constructors that don't take in `nanValueCounts` maps: #2044
> 
> I suppose when we go to deprecate the old `Metrics` constructors, we can simply pass `null` into the `Metrics` constructor for `nanValueCounts` with ORC in order to keep the current behavior. Or update the various evaluator code paths to ensure that a missing key from this map does not indicate anything special if we choose to pass in an empty map.
> 
> This is all based off of my understanding of ORC's `ColumnStatistics` and the notes in this PR about `upperBound ` and `lowerBound` being `NaN` in ORC if _any_ value in the file is `NaN`, so please correct me if I'm mistaken and somebody sees an easy way to get NaN stats from an ORC file. To me, I fail to see how we can get the count of NaN values in the following code block (though as I mentioned we can always continue to pass in `null` to match the existing behavior, provided the various metrics evaluators are aware that `null` for that Map entry for nullValueCounts doesnt convey much meaning for ORC).
> 
> We could also potentially track NaN values in the various orc writers, like the parquet writers do.
> 
> The relevant code for populating ORC metrics is here:
> 
> https://github.com/apache/iceberg/blob/29915e3522f9a808420a2e16fb1c2cf3ae165a74/orc/src/main/java/org/apache/iceberg/orc/OrcMetrics.java#L100-L172

Thanks for your interests in this PR! You are right that some of this PR will not be working without NaN counter for checking if a field contains partial/all NaN, and it's hard to get ORC `nanValueCounts` from ORC file directly. I think today evaluators should treat missing keys from this map or lack of such map the same way as not having NaN value count information for a given field, or do you think we are not doing this in some places?

Regarding populating NaN counter for ORC files, yes we are going to track NaN values in various ORC writers just as we do for parquet writers, and the change is currently in review in #1790.

**Comment by yyanyy** [comment]
I have updated the PR based on comments. Thank you everyone for your interests in this PR!

**Comment by rdblue** [comment]
Thank you for fixing this, @yyanyy! Great work.

Thanks to @jbapple, @jackye1995, and @kbendick for reviewing!
